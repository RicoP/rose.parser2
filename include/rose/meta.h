#pragma once

// this macro will be ignored my the compiler but serves as a 
// placeholder for meta information for the rose header parser.
#define $(...)

struct RIdentity {};

static inline RIdentity identity;