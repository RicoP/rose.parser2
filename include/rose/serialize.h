#pragma once

#include <rose/type.h>

enum class ESerializeStatus : int {
    OK = 0
};

struct RSerializeResult {
    ESerializeStatus status = ESerializeStatus::OK;
    int line = -1;

    static RSerializeResult ok(int line) {
        RSerializeResult result { ESerializeStatus::OK, line };
        return result;
    }
};

struct RSerilalizer {
    virtual ~RSerilalizer() {}
    virtual RSerializeResult serialize_float(float arr, const RTypeInfo &) = 0;
    virtual RSerializeResult serialize_span(float * arr, int size, const RTypeInfo &) = 0;
};

inline RSerializeResult serialize(float * arr, int length, RSerilalizer & serializer, const RTypeInfo & typeinfo) {
    return serializer.serialize_span(arr, length, typeinfo);
}

template<class T, int N>
inline RSerializeResult serialize(T (&arr)[N], RSerilalizer & serializer, const RTypeInfo & typeinfo) {
    return serialize(arr, serializer, typeinfo);
}

template<class T, int N>
inline RSerializeResult serialize(T (&arr)[N], RSerilalizer & serializer) {
    return serialize(arr, serializer, rose_typeof(arr));
}

template<class T>
inline RSerializeResult serialize(T & value, RSerilalizer & serializer) {
    return serialize(value, serializer, rose_typeof(value));
}
