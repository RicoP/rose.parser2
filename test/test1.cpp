#include "test1.h"

RSerializeResult foo(Vector3 & v, RSerilalizer & serializer) {
    return serialize(v, serializer);
}

/*
RSerializeResult bar(Matrix44 & m, RSerilalizer & serializer) {
    return serialize(m, serializer);
}
*/

int main() {
    Vector3 v;
    return 0;
}