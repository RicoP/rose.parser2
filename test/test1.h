#pragma once

#include <rose/meta.h>
#include <rose/serialize.h>

$(foo_t {})
struct Vector3 {
    float x,y,z;
};

inline RSerializeResult serialize(Vector3 & v, RSerilalizer & serializer) {
    float arr[3] = {v.x, v.y, v.z};
    return serialize(arr, serializer);
}

struct Matrix44 {
    float m11,m12,m13,m14;
    float m21,m22,m23,m24;
    float m31,m32,m33,m34;
    float m41,m42,m43,m44;

    Matrix44(RIdentity) {
        m11 = 1; m12 = 0; m13 = 0; m14 = 0;
        m21 = 0; m22 = 1; m23 = 0; m24 = 0;
        m31 = 0; m32 = 0; m33 = 1; m34 = 0;
        m41 = 0; m42 = 0; m43 = 0; m44 = 1;
    }
};

inline RSerializeResult serialize(Matrix44 & mat, RSerilalizer & serializer) {
    float arr[4][4] = {
        {mat.m11,mat.m12,mat.m13,mat.m14},
        {mat.m21,mat.m22,mat.m23,mat.m24},
        {mat.m31,mat.m32,mat.m33,mat.m34},
        {mat.m41,mat.m42,mat.m43,mat.m44}
    };
    return serialize(arr, serializer);
}

struct Hero {
    Matrix44 transform = identity;
    Vector3 location;
    float life = 0;
};
