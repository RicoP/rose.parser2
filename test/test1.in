inline static int foobar1(int n) ;

inline static int foobar2(int n) {
  hello();
}

inline static int foobar3(int n) {
  if (n == 0) { hello(); } 
}

inline static int foobar4(int n) {
  if (n == 0) { 
    if (n == 0) {  
      hello(); 
    }
  } 
}

inline static int foobar5(int n) {
  if (n == 0) { 
      hello(); 
  } 
  if (n == 1) { 
      ballo(); 
  } 
}


inline static int foobar5(int n) {
  if (n == 0) { 
      hello(); 
      hello(); 
  } 
  if (n == 1) { 
      ballo(); 
  } 
}


inline static int foobar6(int n) {
  { 
      hello(); 
      hello(); 
  } 
  if (n == 1) { 
      ballo(); 
  } 
}

inline static int foobar7(int n) {
  { 
      hello(); 
      hello(); 
  } 
  { 
      ballo(); 
  } 
}

inline static int foobar8(int n) {
  { 
      hello(); 
      hello(); 
  } 
  hello(); 

  if (n == 1) { 
      ballo(); 
  } 
}


#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#endif

#include "rose/core/hash.h"

#include "stdio.h"
#include <math.h>

struct Vec2 {
  int x;
  int y;
};

struct Vec3 {
  float x,y = 0,z;
};

//
// Hello World

void A();
void foo();
void baz() {}

inline static int fib(int n) {
  if (n == 0) { return 0; } 
  if (n == 1) { return 1; }
  return fib(n - 1) + fib(n - 2);
}

//
// Hallo Ballo

/* 

// whatever
Multi line comment 

*/

int main() {
  char c = 'c';
  char c2 = 'rico';
  int n;
  int i = 0;
  
  while (i < 10) {
    n = fib(10);
    print(n);
    i = i + 1;
  }

  do {
    print("Hi!");    
  } while(false) ;

  int x = -42;
  float y = 3.14f;
  double z = .001;

  x = y;
  int i = k;
  int k = i+k-k / 42;

  if(x < y) {  print("Hi!");  }
  if(y >= x);
  if(y >= x) {return 0; }

  if(0 < y) {  print("Hi!");  }
  if(y >= 0);
  if(1 >= 0) {return 0; }

  if(x) { return 0; }
  if(!x) { return 0; }
  if(!!x) { return 0; }


  return 0;
}

$(function_t { .x = 4, .y = "hello", w = 4, hello })
inline static int foobarX(int n /* a value / or something else */) {
  hello();
}



//$(record_t {  })
struct Vector3 {
  private:
    int x,y,z;
};
