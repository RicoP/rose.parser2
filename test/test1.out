> 
  regex
  includes|>
    string:1:1 '#include'        
    string|regex:1:10 '"stdio.h"'
  procedure|>
    string:3:1 'int'
    ident|regex:3:5 'fib'        
    char:3:8 '('
    args|typeident|>
      string:3:9 'int'
      ident|regex:3:13 'n'       
    char:3:14 ')'
    body|>
      char:3:16 '{'
      stmt|>
        string:4:3 'if'
        char:4:6 '('
        exp|>
          lexp|term|factor|ident|regex:4:7 'n'
          string:4:9 '=='
          lexp|term|factor|number|regex:4:12 '0'
        char:4:13 ')'
        stmt|>
          char:4:15 '{'
          stmt|>
            string:4:17 'return'
            lexp|term|factor|number|regex:4:24 '0'
            char:4:25 ';'
          char:4:27 '}'
      stmt|>
        string:5:3 'if'
        char:5:6 '('
        exp|>
          lexp|term|factor|ident|regex:5:7 'n'
          string:5:9 '=='
          lexp|term|factor|number|regex:5:12 '1'
        char:5:13 ')'
        stmt|>
          char:5:15 '{'
          stmt|>
            string:5:17 'return'
            lexp|term|factor|number|regex:5:24 '1'
            char:5:25 ';'
          char:5:27 '}'
      stmt|>
        string:6:3 'return'
        lexp|>
          term|factor|>
            ident|regex:6:10 'fib'
            char:6:13 '('
            lexp|>
              term|factor|ident|regex:6:14 'n'
              char:6:16 '-'
              term|factor|number|regex:6:18 '1'
            char:6:19 ')'
          char:6:21 '+'
          term|factor|>
            ident|regex:6:23 'fib'
            char:6:26 '('
            lexp|>
              term|factor|ident|regex:6:27 'n'
              char:6:29 '-'
              term|factor|number|regex:6:31 '2'
            char:6:32 ')'
        char:6:33 ';'
      char:7:1 '}'
  main|>
    string:9:1 'main'
    char:9:5 '('
    char:9:6 ')'
    body|>
      char:9:8 '{'
      decls|>
        typeident|>
          string:11:3 'int'
          ident|regex:11:7 'n'
        char:11:8 ';'
        typeident|>
          string:12:3 'int'
          ident|regex:12:7 'i'
        char:12:8 ';'
      stmt|>
        string:14:3 'while'
        char:14:9 '('
        exp|>
          lexp|term|factor|ident|regex:14:10 'i'
          char:14:12 '<'
          lexp|term|factor|number|regex:14:14 '10'
        char:14:16 ')'
        stmt|>
          char:14:18 '{'
          stmt|>
            ident|regex:15:5 'n'
            char:15:7 '='
            lexp|term|factor|>
              ident|regex:15:9 'fib'
              char:15:12 '('
              lexp|term|factor|number|regex:15:13 '10'
              char:15:15 ')'
            char:15:16 ';'
          stmt|>
            string:16:5 'print'
            char:16:10 '('
            lexp|term|factor|ident|regex:16:11 'n'
            char:16:12 ')'
            char:16:13 ';'
          stmt|>
            ident|regex:17:5 'i'
            char:17:7 '='
            lexp|>
              term|factor|ident|regex:17:9 'i'
              char:17:11 '+'
              term|factor|number|regex:17:13 '1'
            char:17:14 ';'
          char:18:3 '}'
      stmt|>
        string:20:3 'return'
        lexp|term|factor|number|regex:20:10 '0'
        char:20:11 ';'
      char:21:1 '}'
  regex
